components/ee/ws-scheduler/pkg/scheduler/state.go

```diff
@@ -67,56 +68,112 @@ type ResourceUsage struct {
 	UsedOther *res.Quantity
 }
 
-// NewState creates a fresh, clean state
-func NewState() *State {
-	return &State{
-		Nodes: make(map[string]*Node),
-		Pods:  make(map[string]*corev1.Pod),
+func newResourceUsage(total *res.Quantity) ResourceUsage {
+	return ResourceUsage{
+		Total:        total.Copy(),
+		Available:    res.NewQuantity(0, res.BinarySI),
+		UsedHeadless: res.NewQuantity(0, res.BinarySI),
+		UsedOther:    res.NewQuantity(0, res.BinarySI),
+		UsedRegular:  res.NewQuantity(0, res.BinarySI),
 	}
 }
 
-// UpdateNodes integrates the given node into the current state
-func (s *State) UpdateNodes(nodes []*corev1.Node) {
-	for _, node := range nodes {
-		s.UpdateNode(node)
-	}
+func (r *ResourceUsage) updateAvailable() {
+	r.Available = r.Total.Copy()
+	r.Available.Sub(*r.UsedHeadless)
+	r.Available.Sub(*r.UsedOther)
+	r.Available.Sub(*r.UsedRegular)
 }
 
-// UpdateNode integrates the given node into the current state
-func (s *State) UpdateNode(node *corev1.Node) {
-	n := s.Nodes[node.Name]
-	if n == nil {
-		n = createNodeFrom(node)
-		s.Nodes[node.Name] = n
-	} else {
-		n.update(node)
+// ComputeState builds a new state based on the current world view
+func ComputeState(nodes []*corev1.Node, pods []*corev1.Pod, bindings []*Binding) *State {
+	var (
+		nds       = make(map[string]*Node)
+		pds       = make(map[string]*corev1.Pod)
+		podToNode = make(map[string]map[string]struct{}, len(nodes))
+	)
+
+	for _, n := range nodes {
+		nds[n.Name] = &Node{
+			Node:     n,
+			Services: make(map[string]struct{}),
+		}
+		podToNode[n.Name] = make(map[string]struct{})
 	}
+	for _, p := range pods {
+		pds[p.Name] = p
 
-	s.updateAssignedPods(n)
-}
+		if p.Spec.NodeName == "" {
+			continue
+		}
+		ptn, ok := podToNode[p.Spec.NodeName]
+		if !ok {
+			log.WithField("podName", p.Name).WithField("nodeName", p.Spec.NodeName).Warn("pod is bound to unknown node")
+			continue
+		}
+		ptn[p.Name] = struct{}{}
+	}
+	for _, b := range bindings {
```

Comment:
A short comment stating that local bindings take precedence and we assume that they are more up-to-date would be nice.